### PR TITLE
Don't cache Angle::Unit

### PR DESCRIPTION
**Pull request by wjp**
Profiling shows the uncached version is twice as fast. It also saves a megabyte of regularly accessed memory which should relieve some cache pressure.

### CODE REVIEW

File: source/Angle.cpp [hunk 1]
```diff
@@ -121,22 +121,7 @@ Angle Angle::operator-() const
...
-	if(cache.empty())
-	{
-		cache.reserve(STEPS);
-		for(int i = 0; i < STEPS; ++i)
-		{
-			double radians = i * STEP_TO_RAD;
-			// The graphics use the usual screen coordinate system, meaning that
-			// positive Y is down rather than up. Angles are clock angles, i.e.
-			// 0 is 12:00 and angles increase in the clockwise direction. So, an
-			// angle of 0 degrees is pointing in the direction (0, -1).
```
Comment (quyykk) [inline source/Angle.cpp:135]: I think we should keep this comment

**Review by quyykk — COMMENTED** [review]
> Profiling shows the uncached version is twice as fast

oh wow.

File: source/Angle.cpp (hunk 1 above)
Comment (wjp) [reply source/Angle.cpp:135]: Ah, good point. That was a bit of an overzealous delete.

**Comment by Amazinite** [comment]
> > Profiling shows the uncached version is twice as fast
>
> oh wow.

How, though?

**Comment by wjp** [comment]
I just ran a quick test, and computing a billion (sin,cos) pairs takes 40s here, so that's 40ns each. Reading from the cache a billion times in a row in an isolated benchmark is quite a bit faster than that (but of course that's straight from the CPU cache), so that's an interesting question. I'll do some more profiling runs of less isolated calls tomorrow (or later this week depending on time) to see if I can narrow this down a bit further.

**Comment by mike-f1** [comment]
[Relevant link](https://stackoverflow.com/questions/12333638/how-much-cycles-math-functions-take-on-modern-processors) for time took by sin&cos, and a [relevant link](https://youtu.be/EAUlxpdj3fY?t=346) for time to fetch data (tough I saw different times, like L1 -> 5; L2 -> 20; L3 -> 50 but I think is processor dependent...)

> How, though?

Probably because any call to `Angle::Unit()` imply an entire cache line loaded 'till CPU (64bytes to read 1*sizeof(Point)).
And even if that's not (probably, but I would not bet on this either...) so slow per se, it _unload_ cached data for other purposes with a big lever: you unload 64bytes (or 32bytes to 256bytes, according to the second link) to read a single Point.

Then a good question to ask is: how likely another read of an angle can be in the same cache line?
Well: you have 2^16 angles, and no ordering. So there's no guarantee that you will load from a low cache, even if you have a massive combat between two fleets approaching (that's the "ordering" thing). And we are speaking of 1MBytes of data, while on my old steam-powered-processor (an `i5-3210M`), cache sizes are:
```
L1d cache:                       64 KiB
L1i cache:                       64 KiB
L2 cache:                        512 KiB
L3 cache:                        3 MiB
```

Thus, I would answer "unlikely".

I think (probably I should have said "bet") the vast majority of angles still in memory, while some reside in L3
and approach CPU from there only when requested (so ~21 to ~200 cycles averaged) -> roughly the same
of computing sin&cos. 

> I'll do some more profiling runs of less isolated calls tomorrow

Why you don't profile `AI::AutoFire`: it seems one of the best scenario where `Angle`s can play a role... Or show your already obtained profiled data.

NOTE: Wouldn't this PR be worth to be mentioned in "Analyzing performance bottlenecks" issue (6329)?

**Comment by Amazinite** [comment]
> I'll do some more profiling runs of less isolated calls tomorrow (or later this week depending on time) to see if I can narrow this down a bit further.

@wjp You ever get around to that, or able to get to it soon?

**Comment by tibetiroka** [comment]
I did some testing on the performance of this PR. I wouldn't call it a proper benchmark, but it does work pretty good as a stress test.

## Setup
This save file ([Test Pilot~'olofez benchmark.txt](https://github.com/endless-sky/endless-sky/files/11938535/Test.Pilot.olofez.benchmark.txt)) contains a flagship with 10k other ships. The test is to measure how long it takes (in real time) to take off from a planet and land on it again immediately.

I tested master (f83b859) against this PR merged with the same master.


## Results

|run #| master |   7769   |
|--------|------------|------------|
| 1      | 3:56.89 | 3:45.76 |
| 2      | 3:44.56 | 4:00.41 |
| 3      | 3:53.87 | 3:57.57 |

It doesn't look like there is much difference in performance. I'd say the differences between the two results fall within the margin of error.

**Comment by warp-core** [comment]
What processor do you have? And, what speed is your system memory?

I would expect caching to be slower in systems without a whole lot of CPU cache.

Also, did you restart the game between each run?

**Comment by tibetiroka** [comment]
CPU: AMD Ryzen 7 3700x
RAM: 2x8 GB 3200 MT/s DDR4
Kernel: Linux 6.4.1-zen1-1-zen
(I was using the `powersave` governor at the time.)

I didn't restart the game between these runs, but I did several tests before these to figure out how to best use this test, and that included multiple restarts. Those initial results were in the same range as these ones.

**Comment by Amazinite** [comment]
A problem is that there's a lot more calculations going on than just angle calculations when in flight. At best we can say that if this is a performance improvement to angles, angles themselves aren't a significant enough contributor to the calculations any given frame to make enough of a difference.

Since we aren't sure if this is an actual improvement or not and it's just been sitting here, I'll be closing this. If at any point someone can show profiling results that suggest that this actually is an improvement, then we can reopen it and reconsider.

**Pull request closed without merging**